### ISSUE TITLE
cloudinit does not find private ip address on rackspace onmetal servers

### ISSUE THREAD

**Issue by crawford** [coreos/bugs#757]
<a href="https://github.com/paulczar"><img src="https://avatars.githubusercontent.com/u/2488346?v=3" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [paulczar](https://github.com/paulczar)**
_Saturday Aug 02, 2014 at 17:04 GMT_
_Originally opened as https://github.com/coreos/coreos-cloudinit/issues/195_

---

coreos on RAX onmetal server doesn't get private_ip from metadata ...   coreos on RAX vm does.

coreos-cloudinit on the onmetal server is version `0.9.0`, on the VM its `0.8.9`.

userdata used for both 

```
#cloud-config

coreos:
  etcd:
    # generate a new token for each unique cluster from https://discovery.etcd.io/new
    discovery: https://discovery.etcd.io/22575831156ee709af2d6e9969a59885
    # multi-region and multi-cloud deployments need to use $public_ipv4
    addr: $private_ipv4:4001
    peer-addr: $private_ipv4:7001
  units:
    - name: etcd.service
      command: start
    - name: fleet.service
      command: start
```

etcd service

```
$ cat /run/systemd/system/etcd.service.d/20-cloudinit.conf
[Service]
Environment="ETCD_ADDR=:4001"
Environment="ETCD_DISCOVERY=https://discovery.etcd.io/9d2e01fdd2b339ac83a5a3bbd2cae19d"
Environment="ETCD_NAME=a60ee84b68f0436f8aacdec19ff51106"
Environment="ETCD_PEER_ADDR=:7001"
```

openstack metadata from config-drive

```
$ cat openstack/latest/meta_data.json 
{"admin_pass": "cs32b49aFLnQ", "random_seed": "EMPUtmxyfpD4kM0Db0Jxdr7X9P5IsQ8d38cGAW2F8QdHdfZCxc2rebCFCZe393wD6iHibVFSqWu+vc3ORnM5WtK7eI552NeKujAaLKBi7HVc+DY8pwMaOjTJ7PK7Xl27+RE1u4CyywhFutpHu+9Yp/0RTSOQ0g1Cv43O2kuJCCrXbzh+HmuKGYUSOHH/bFu9w5BzaNIqq61cXVe+YssOpbfNMdgDHDThAyK1DA0Qvm4KDCehcrO9CsZXKuQ2h7v2GSbWXF5IMNL/+U7jMtD3g7VKVQ6mitiQZrvhTxVg2MCss25mmLcQRYFXbEev0DOsfgG4OjjrSX2DAcr1M8FlX7aC5qqcJrSGyvZUEyTMYqQKML6r2gBLGEvyACwoCkzS1Eqx8D70UCkUENk5xK/l7hCr+jYHBTqIiSTFTMfdwzQdL7fowE3V/epcsrxDVxFXkNrC10holkJ6ckqKbHG7tKcBXuO0n5jd1nC7GRuBC7WJhbHBHI+HVv2T7xpWmLUkpBCl2fYYQ8Ay03rF378UNaOYxb/M8CdsTIdfXRv/wyxTD6PdlxK3wqlWx4Rxr+Rtn8jgupBzgicqSeIWLgM/8HcTWGyQMR2Vv0Wqk4recDjvzQv6WUNIYOhjZP9XcufPQV2D4ZeB2WvSt5BhKyMDLL7Qdnw+Ugo+KkBV+lyrlRw=", "name": "DEIS_Server_OnMetal", "availability_zone": "nova", "hostname": "deis-server-onmetal.novalocal", "launch_index": 0, "public_keys": {"XXX": "ssh-rsa XXXXXt\n"}, "network_config": {"content_path": "/content/0000", "name": "network_config"}, "uuid": "de1a7baf-e527-4a6f-9573-6c8822abb6a5"}
```

ec2 metadata from config-drive

```
$ cat ec2/latest/meta-data.json
{"reservation-id": "r-wiplq437", "hostname": "deis-server-onmetal.novalocal", "security-groups": ["default"], "ramdisk-id": "ari-00000002", "public-ipv4": "", "ami-manifest-path": "FIXME", "instance-type": "OnMetal Compute v1", "instance-id": "i-0000110a", "local-ipv4": "104.130.21.133", "local-hostname": "deis-server-onmetal.novalocal", "placement": {"availability-zone": "nova"}, "ami-launch-index": 0, "public-hostname": "deis-server-onmetal.novalocal", "public-keys": {"0": {"openssh-key": "XXXX", "_name": "0=XXXX"}}, "ami-id": "ami-00000025", "kernel-id": "aki-00000002", "instance-action": "none", "block-device-mapping": {"ami": "sda", "root": "/dev/sda"}}
```

**Comment by crawford** [comment]
<a href="https://github.com/crawford"><img src="https://avatars.githubusercontent.com/u/368848?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [crawford](https://github.com/crawford)**
_Saturday Aug 02, 2014 at 17:34 GMT_

----

Unfortunately, this is an issue with OnMetal at the moment. As you can see in the metadata, `public-ipv4` is empty and `local-ipv4` is actually your public address. Once their metadata is corrected, https://github.com/coreos/coreos-cloudinit/pull/188 will fix the issue.

As a workaround, you can populate `/etc/environment` with the addresses for your instance.
```sh
COREOS_PUBLIC_IPV4=1.2.3.4
COREOS_PRIVATE_IPV4=5.6.7.8
```

**Comment by crawford** [comment]
<a href="https://github.com/paulczar"><img src="https://avatars.githubusercontent.com/u/2488346?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [paulczar](https://github.com/paulczar)**
_Saturday Aug 02, 2014 at 22:52 GMT_

----

this appears to work-around the issue ...  

```
#cloud-config
---
coreos:
  etcd:
    # generate a new token for each unique cluster from https://discovery.etcd.io/new
    # uncomment the following line and replace it with your discovery URL
    addr: $private_ipv4:4001
    peer-addr: $private_ipv4:7001    
    discovery: https://discovery.etcd.io/7cbd29a28b800385c89b7781b0335d42
  fleet:
  units:
  - name: etcd.service
    command: start
    after: create-etcd-env.service
  - name: fleet.service
    command: start
    after: etcd.service
  - name: create-etcd-env.service
    command: start
    content: |
      [Unit]
      Description=creates etcd environment

      [Service]
      Before=etcd.service
      Type=oneshot
      ExecStart=/bin/sh -c "sed -i \"s/=:/=`ifconfig bond0.401 | grep 'inet ' | awk '{print $2}'`:/\" /run/systemd/system/etcd.service.d/20-cloudinit.conf && systemctl daemon-reload && systemctl etcd restart"
```

**Comment by crawford** [comment]
<a href="https://github.com/philips"><img src="https://avatars.githubusercontent.com/u/33544?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [philips](https://github.com/philips)**
_Monday Aug 04, 2014 at 16:22 GMT_

----

@crawford is there an ETA from the Rackspace OnMetal team for fixing this?

**Comment by crawford** [comment]
<a href="https://github.com/crawford"><img src="https://avatars.githubusercontent.com/u/368848?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [crawford](https://github.com/crawford)**
_Tuesday Aug 05, 2014 at 16:57 GMT_

----

@philips I haven't heard back from them but I wouldn't hold your breath. We are investigating a solution that involves a modification to etcd.

**Comment by crawford** [comment]
<a href="https://github.com/chrissnell"><img src="https://avatars.githubusercontent.com/u/1072626?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [chrissnell](https://github.com/chrissnell)**
_Thursday Aug 28, 2014 at 19:04 GMT_

----

Is there a reason that we couldn't have a $local_ipv4 substitution in cloud-config?   I'm having the same problem but we're on the Rackspace Private Cloud product and I think that it's going to be a major uphill battle to change this.    On my Ubuntu images, I wrote a run-once-at-firstboot script that pulls local-ipv4 from http://169.254.169.254/2009-04-04/meta-data/local-ipv4 and sets my stuff up appropriately.   

I would like to use $local_ipv4 in CoreOS.   

Snelly

**Comment by crawford** [comment]
<a href="https://github.com/chrissnell"><img src="https://avatars.githubusercontent.com/u/1072626?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [chrissnell](https://github.com/chrissnell)**
_Thursday Aug 28, 2014 at 19:26 GMT_

----

Disregard my last comment.  Digging into the source, I now see that local-ipv4 is being fetched.  For some reason, the substitution in my cloud-config fails.  I'm not sure why, but that is a separate issue and not related to this issue.

**Comment by crawford** [comment]
<a href="https://github.com/jayofdoom"><img src="https://avatars.githubusercontent.com/u/1738052?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [jayofdoom](https://github.com/jayofdoom)**
_Thursday Aug 28, 2014 at 19:35 GMT_

----

@crawford We now including the network information via json format in cloud-config vendor_data.json. This can probably be utilized to help resolve this bug.

**Comment by crawford** [comment]
We are no longer working on coreos-cloudinit. If this functionality is still desired, it can be added to coreos-metadata.
